pennylane/qinfo/transforms.py

```diff
@@ -58,11 +58,14 @@ def wrapper(*args, **kwargs):
         if len(measurements) != 1 or not isinstance(measurements[0], StateMP):
             raise ValueError("The qfunc measurement needs to be State.")
 
+        dm_measurement = measurements[0].wires or "mixed" in qnode.device.name
```

Comment:
Linking the discussion from https://github.com/PennyLaneAI/pennylane/pull/4168#discussion_r1204728774:

@albi3ro says:
> Separating `StateMP` from `DensityMatrixMP` would definitely affect this line.
> 
> It'd be especially handy if `StateMP` always returned a state and a `DensityMatrixMP` always returned a density matrix without being device dependent.

@eddddddy says:
> Yep, I think we'd all prefer that. Since the separation is in prototyping right now, I assume it won't be merged anytime soon? I can add a TODO here mentioning that this would need updating once it's done.

Reply:
I'm almost scared to say this now, but.. if we introduced a `DensityMatrixMP` before doing any of this, could we propagate the type along such that we can keep the dispatching in `reduced_dm` according to that measurement type?

Reply:
Unfortunately, I think we want to keep `qml.math` separate from the device/measurements parts of PennyLane. That means no inclusion of any measurement logic whatsoever in `qml.math`.

Reply:
I agree with @eddddddy . Separating the logic means that we don't need to dispatch anyway.

Reply:
I think the separation of `StateMP` and `DensityMatrixMP` is way out of scope of this PR and even this whole mini-project, recalling that the aim here is just to support broadcasting in `qinfo`. I did add a TODO comment here that addresses that issue.